sdk/helper/scaleutils/nodeselector/empty_nodes.go

```diff
@@ -0,0 +1,95 @@
+package nodeselector
+
+import (
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad-autoscaler/sdk"
+	"github.com/hashicorp/nomad/api"
+)
+
+// emptyNodesClusterScaleInNodeSelector is the NodeSelector implementation of
+// the ClusterScaleInNodeSelector interface. It selects nodes for termination
+// only if they do not have any non-terminal allocations, and therefore can be
+// classed as empty.
+type emptyNodesClusterScaleInNodeSelector struct {
+	client *api.Client
+	log    hclog.Logger
+}
+
+// newEmptyNodesClusterScaleInNodeSelector returns a new
+// emptyNodesClusterScaleInNodeSelector implementation of the
+// ClusterScaleInNodeSelector interface.
+func newEmptyNodesClusterScaleInNodeSelector(c *api.Client, log hclog.Logger) ClusterScaleInNodeSelector {
+	return &emptyNodesClusterScaleInNodeSelector{
+		client: c,
+		log:    log,
+	}
+}
+
+// Name satisfies the Name function on the ClusterScaleInNodeSelector
+// interface.
+func (e *emptyNodesClusterScaleInNodeSelector) Name() string {
+	return sdk.TargetNodeSelectorStrategyEmptyNodes
+}
+
+// Select satisfies the Select function on the ClusterScaleInNodeSelector
+// interface.
+func (e *emptyNodesClusterScaleInNodeSelector) Select(nodes []*api.NodeListStub, num int) []*api.NodeListStub {
+
+	var out []*api.NodeListStub
+
+	for _, node := range nodes {
+
+		// If the node has non-terminal allocations, this isn't the node we are
+		// looking for. Otherwise add this node to our selected list.
+		if e.nodeHasAllocs(node.ID) {
+			continue
+		}
+		out = append(out, node)
+
+		// Short circuit if we have found a required number of nodes. We do not
+		// need to do any further filtering.
+		if len(out) == num {
+			break
+		}
+	}
+
+	return out
+}
+
+// nodeHasAllocs returns whether the node has non-terminal allocations.
+func (e *emptyNodesClusterScaleInNodeSelector) nodeHasAllocs(id string) bool {
+
+	allocs, _, err := e.client.Nodes().Allocations(id, nil)
+	if err != nil {
+		e.log.Error("failed to detail node allocs", "error", err)
+		return true
+	}
+
+	for _, alloc := range allocs {
+		if !allocClientTerminalStatus(alloc) || !allocServerTerminalStatus(alloc) {
+			e.log.Debug("selector skipped node with running allocs", "node_id", id)
+			return true
+		}
+	}
+	return false
+}
+
+// TODO (jrasell) remove once Nomad#10230 is released.
```

Comment:
is release necessary or is merged sufficient?

Reply:
This depends if we are happy to upgrade the API dependancy to main, rather than using the release SHA?

Reply:
Personally I wouldn't mind using a non-release SHA if there's something useful to us 👍